### PR TITLE
Build M-series (arm64) macOS wheels on GitHub Actions hosted runners

### LINKED ISSUE #3772: Use self-hosted macOS M2 runner to build macOS arm64 or universal wheels
@BradyPlanden and I, in conversation about the usage of the runner for the repository – have decided to look into building wheels for M-series platforms starting with PyBaMM v24.5rcX. This is since cross-compilation for SuiteSparse requires at least a version of 7.2.1, and we support 6.0.3, so we might not be able to do that yet on the standard GitHub Actions runners (see #3462).

In this scenario, this requires setting up a job in the PyPI workflows where the wheels can be built and passed as artifacts to the job that uploads them.

Whether to build a universal macOS wheel or an architecture-specific wheel can be discussed based on the wheel tags we are able to achieve, the macOS X version constraints for such wheels, and the size of the binaries generated.

### PR DESCRIPTION
**Pull request by agriyakhetarpal**
# Description

GitHub Actions M1 runners are free for open-source repositories now – I have configured the PyPI workflow to add a job for these runners.

> [!IMPORTANT]
> The wheels can be tested on M-series hardware via the following link: https://github.com/pybamm-team/PyBaMM/actions/runs/7721068374/

Closes #3772, related to #3462

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [x] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
Note to self:

- [ ] Look into if archflags configuration is needed for CMake
- [ ] Add CHANGELOG entry
- [ ] Investigate whether to build universal2 wheel via `delocate-fuse`, `delocate-addplat`, or similar or to keep macOS wheels architecture-specific

**Comment by agriyakhetarpal** [comment]
We should start testing PyBaMM on M1 runners as well if we are going to be distributing them – I can do that in a follow-up PR, though I am not sure if #3140 (particle cracking submodels) has been resolved or not.

File: .github/workflows/publish_pypi.yml
Comment (Saransh-cpp) [inline .github/workflows/publish_pypi.yml:250]: Why was `always()` removed from here?

**Review by Saransh-cpp — APPROVED** [review]
Thanks, @agriyakhetarpal! This looks really good!

File: .github/workflows/publish_pypi.yml
Comment (agriyakhetarpal) [reply .github/workflows/publish_pypi.yml:250]: I removed it because its presence negated the existence of the other conditions – this job ran even in forks despite not being supposed to. The check for failures in the condition seems to be good enough for this purpose.

Comment (Saransh-cpp) [reply .github/workflows/publish_pypi.yml:250]: This would not work without `always()`. If any of the wheel jobs fail, the `open_failure_issue` job will never be executed. Using always is a bit hacky but that is what makes this work.

Maybe we can create a new issue for this and keep it as it is until we find a solution?

**Comment by Saransh-cpp** [comment]
Also, all green ticks in this PR, nice!

File: .github/workflows/publish_pypi.yml
Comment (agriyakhetarpal) [reply .github/workflows/publish_pypi.yml:250]: Makes sense, Actions are disabled by default on forks anyway. Happy to revert the change for now

Comment (agriyakhetarpal) [inline .github/workflows/publish_pypi.yml:250]:
```suggestion
    if: ${{ always() && contains(needs.*.result, 'failure') && github.repository_owner == 'pybamm-team'}}
```

**Review by Saransh-cpp — APPROVED** [review]

**Comment by kratman** [comment]
> I am not sure if #3140 (particle cracking submodels) has been resolved or not.

I put in a stop-gap fix for that ticket. I did not mark it as closed because all I did was adjust the grid. Not really a full fix or investigation. Having tests that would alert us sooner would probably make fixing stuff like that more visible and we could discuss closing that issue if we start testing more thoroughly on M-series hardware.

**Comment by agriyakhetarpal** [comment]
> Having tests that would alert us sooner would probably make fixing stuff like that more visible and we could discuss closing that issue if we start testing more thoroughly on M-series hardware.

Sure, happy to merge this after you can download and test a wheel or two (I tested them, they are working for me since the `IDAKLUSolver` can be instantiated and can be used to solve a basic Simulation – even though that seems to not be a reliable test, see #3783).

I will set up testing against the M1 runners in a follow-up PR as soon after this is merged.

**Comment by kratman** [comment]
@agriyakhetarpal I downloaded the wheels to test them

**Review by kratman — APPROVED** [review]

**Pull request merged**
